docfx_yaml/extension.py

```diff
@@ -1417,58 +1417,127 @@ def find_markdown_pages(app, outdir):
                 'href': mdfile.name.lower(),
             })
 
+def resolve_cross_reference(
+    words: List[str],
+    index: int,
+    current_word: str,
+    keyword: str,
+    visited_words: List[str],
+    hard_coded_references: Dict[str, str] = None
+) -> Optional[str]:
+    """Given a word, check if it contains a reference keyword we should convert.
 
-# Finds and replaces occurrences which should be a cross reference in the given
-# content, except for the current name.
-def convert_cross_references(content: str, current_name: str, entry_names: List[str]):
+    Args:
+        words: list of words used to check and compare content before and after `current_word`
+        index: index position of `current_word` within words
+        current_word: current word being looked at
+        keyword: reference keyword to check against
+        visited_words: list of words containing words that's been processed so far
+        hard_coded_references: Optional list containing a list of hard coded reference
+
+    Returns:
+        None if current word does not contain a reference keyword, or a string
+          that contains the converted reference.
+    """
+    if keyword in current_word:
+        if hard_coded_references and keyword in hard_coded_references:
+            if "<a" not in words[index-1] and \
+                not (visited_words and f"<a href=\"{hard_coded_references[keyword]}" in visited_words[-1]):
+                return f"<a href=\"{hard_coded_references[keyword]}\">{keyword}</a>"
+
+        else:
+            # If it is already processed as cross reference, skip over it.
+            if "<xref" not in words[index-1] and \
+                not (visited_words and f"<xref uid=\"{keyword}" in visited_words[-1]):
+                return f"<xref uid=\"{keyword}\">{keyword}</xref>"
+
+    return None
+
+
+def convert_cross_references(content: str, current_object_name: str, entry_names: List[str]):
+    """Finds and replaces references that should be a cross reference in given content.
+
+    This should not convert any references that contain `current_object_name`,
+    i.e. if we're processing docstring for `google.cloud.spanner.v1.services`,
+    references to `google.cloud.spanner.v1.services` should not be convereted
+    to references.
+
+    Args:
+        content: given body of content to parse and look for references
+        current_object_name: the name of the current Python object being looked at
+        entry_names: list of keyword references to look for
+
+    Returns:
+        content that has been modified with proper cross references if found.
+
+    """
     words = content.split(" ")
-    new_words = []
+
+    # Contains a list of words that is not a valid reference or converted
+    # references.
+    visited_words = []
```

Comment:
What about `processed`, `resolved`, or `converted` `content`? Or even `result`?

(Visited makes me think of doing a graph traversal.)